quickwit-serve/src/test_utils/cluster_sandbox.rs

```diff
@@ -0,0 +1,255 @@
+            Result::<_, anyhow::Error>::Ok(())
+        });
+        wait_for_server_ready(node_config.quickwit_config.grpc_listen_addr).await?;
+        let mut grpc_search_clients = HashMap::new();
+        let search_client =
+            create_search_service_client(node_config.quickwit_config.grpc_listen_addr).await?;
+        grpc_search_clients.insert(node_config.quickwit_config.grpc_listen_addr, search_client);
+        Ok(Self {
+            node_configs,
+            grpc_search_clients,
+            rest_client: QuickwitRestClient::new(node_config.quickwit_config.rest_listen_addr),
+            index_id_for_test: index_for_test,
+            _temp_dir: temp_dir,
+        })
+    }
+
+    // For now, starts only 3 nodes:
+    // - 2 searchers.
+    // - 1 indexer.
+    // Currently we cannot start several indexers in the same process as they
+    // will share the same `INGEST_API_SERVICE_INSTANCE`.
+    pub async fn start_cluster_nodes() -> anyhow::Result<Self> {
+        let temp_dir = tempfile::tempdir()?;
+        let nodes_services = vec![
+            HashSet::from_iter([QuickwitService::Searcher]),
+            HashSet::from_iter([QuickwitService::Searcher]),
+            HashSet::from_iter([QuickwitService::Indexer]),
+        ];
+        let node_configs = build_node_configs(temp_dir.path().to_path_buf(), nodes_services);
+        let index_for_test = new_coolid("index-for-test");
+        // Creates an index before starting nodes as currently Quickwit does not support
+        // dynamic creation/deletion of indexes/sources.
+        create_index_for_test(&index_for_test, &node_configs[0].quickwit_config).await?;
+        for node_config in node_configs.iter() {
+            let node_config_clone = node_config.clone();
+            tokio::spawn(async move {
+                let result = serve_quickwit(
+                    node_config_clone.quickwit_config,
+                    &node_config_clone.services,
+                )
+                .await;
+                println!("Quickwit server terminated: {:?}", result);
+                Result::<_, anyhow::Error>::Ok(())
+            });
+        }
+        let indexer_config = node_configs
+            .iter()
+            .find(|node_config| node_config.services.contains(&QuickwitService::Indexer))
+            .cloned()
+            .unwrap();
+        let mut grpc_search_clients = HashMap::new();
+        for node_config in node_configs.iter() {
+            if node_config.services.contains(&QuickwitService::Searcher) {
+                continue;
+            }
+            let search_client =
+                create_search_service_client(node_config.quickwit_config.grpc_listen_addr).await?;
+            grpc_search_clients.insert(search_client.grpc_addr(), search_client);
+        }
+        // Wait for a duration greater than chitchat GOSSIP_INTERVAL (50ms) so that the cluster is
+        // formed.
+        tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
That interval might not be enough in CI env where the tests run very slowly. Is there a way we can wait on some condition/state?

Reply:
We can do some polling yes. Let's expose a method to `ClusterSandbox` that is waiting for members.